File: src/K이분탐색/나무자르기.py
import sys
input = sys.stdin.readline

n, m = [int(v) for v in input().split()]
tree = list(map(int, input().split()))

start, end = 0, max(tree)

while start< end:
    mid = (start+end)// 2

    cut = 0
    for i in tree:
        if i > mid:
            cut += i - mid
    
    if cut < m:
        end = mid
    else:
        start = mid + 1

print(start - 1)

File: src/backjun/K이분탐색/나무자르기.java
package backjun.K이분탐색;

import java.util.Scanner;
import java.util.Arrays;
public class 나무자르기 {
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(), m = sc.nextInt();
        int[] tree = new int[n];

        for(int i=0; i<n; i++){
            tree[i] = sc.nextInt();
        }
        sc.close();
        long start = 0, end = Arrays.stream(tree).max().getAsInt()+1;

        long mid = 0;
        while(start < end){
            mid = (start + end) / 2;

            long cut = 0;
            for(long i:tree){
                if(i > mid){
                    cut += i - mid;
                }
            }

            if (cut < m){
                end = mid;
            } else {
                start = mid + 1;
            }
        }

        System.out.println(start -1);
    }
    
}
